Evict unused fonts after 4 rendering updates containing paint changes.

mattwoodrow opened the pull request:
#### 2da5f7dd5dd8835df69e5d489132dfc374c54c52
<pre>
Evict unused fonts after 4 rendering updates containing paint changes.
<a href="https://bugs.webkit.org/show_bug.cgi?id=243189">https://bugs.webkit.org/show_bug.cgi?id=243189</a>

Reviewed by Simon Fraser.

We previously evicted fonts from the resource cache if they hadn&apos;t been used
for 4 rendering updates. Many rendering updates don&apos;t contain any paint changes,
so always count all fonts as &apos;unused&apos;.

This changes us to only count rendering updates that contain painting towards
deciding if a font has been unused long enough to be evicted.

* ../../Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::finalizeRenderingUpdate):
(WebKit::RemoteRenderingBackendProxy::didPaintLayers):
* ../../Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* ../../Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordFontUse):
(WebKit::RemoteResourceCacheProxy::finalizeRenderingUpdateForFonts):
(WebKit::RemoteResourceCacheProxy::didPaintLayers):
(WebKit::RemoteResourceCacheProxy::finalizeRenderingUpdate): Deleted.
* ../../Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:
* ../../Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
* ../../Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didUpdateRendering):
(WebKit::WebPage::didPaintLayers):
* ../../Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/252844@main">https://commits.webkit.org/252844@main</a>
</pre>

smfr commented on Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:
```diff
@@ -87,6 +87,7 @@ class RemoteResourceCacheProxy : public WebCore::NativeImage::Observer, public W
     unsigned m_numberOfFontsUsedInCurrentRenderingUpdate { 0 };
 
     RemoteRenderingBackendProxy& m_remoteRenderingBackendProxy;
+    RenderingUpdateID m_renderingUpdateID;
```
I think it's confusing use a RenderingUpdateID this here when it will have a different value from the RenderingUpdateID in the RemoteRenderingBackendProxy. I think it would be fine to use an `unsigned` count.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
```diff
@@ -378,6 +378,8 @@
     RefPtr<BackingStoreFlusher> backingStoreFlusher = BackingStoreFlusher::create(WebProcess::singleton().parentProcessConnection(), WTFMove(commitEncoder), WTFMove(flushers));
     m_pendingBackingStoreFlusher = backingStoreFlusher;
 
+    m_webPage.didUpdateRendering(false, !!flushers.size());
```
`false` is a boolean trap; I don't know what it means without looking at the source. Maybe used an enum, or just use a separate function?

mattwoodrow committed:
Evict unused fonts after 4 rendering updates containing paint changes.
https://bugs.webkit.org/show_bug.cgi?id=243189

Reviewed by Simon Fraser.

We previously evicted fonts from the resource cache if they hadn't been used
for 4 rendering updates. Many rendering updates don't contain any paint changes,
so always count all fonts as 'unused'.

This changes us to only count rendering updates that contain painting towards
deciding if a font has been unused long enough to be evicted.

* ../../Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::finalizeRenderingUpdate):
(WebKit::RemoteRenderingBackendProxy::didPaintLayers):
* ../../Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* ../../Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordFontUse):
(WebKit::RemoteResourceCacheProxy::finalizeRenderingUpdateForFonts):
(WebKit::RemoteResourceCacheProxy::didPaintLayers):
(WebKit::RemoteResourceCacheProxy::finalizeRenderingUpdate): Deleted.
* ../../Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:
* ../../Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
* ../../Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didUpdateRendering):
(WebKit::WebPage::didPaintLayers):
* ../../Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/252844@main
